pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// imageResolver is an interface used mostly to mock digestResolver for tests.
+type imageResolver interface {
+	Resolve(ctx context.Context, image string, opt k8schain.Options, registriesToSkip sets.String) (string, error)
+}
+
+// backgroundResolver performs background downloads of image digests.
+type backgroundResolver struct {
+	logger *zap.SugaredLogger
+
+	resolver imageResolver
+	enqueue  func(types.NamespacedName)
+
+	queue workqueue.Interface
+
+	mu      sync.Mutex
+	results map[types.NamespacedName]*resolveResult
+}
+
+// resolveResult is the overall result for a particular revision. We create a
+// workItem for each container we need to resolve for the overall result.
+type resolveResult struct {
+	statuses  []v1.ContainerStatus
+	err       error
+	remaining int
+
+	opt              k8schain.Options
+	registriesToSkip sets.String
```

Comment:
This doesn't seem to be set and I'd say we shouldn't set it either. Instead, should we prevent enqueuing the respective image altogether. Is a bit more efficient too I suppose and allows us to move through with a revision that has all images skipped without a wait-for-enqueue.

Reply:
ha good catch - oops, clearly missing a test there. 

Yeah good call, I'll pull the registriesToSkip up in to this and out of digestResolver, and we can skip the queue entirely if it's set.

Reply:
I guess the same reason as with the timeout wrt. config map updatability. Needs to be hooked up tho :yum:

Reply:
fixed this so it passes the params properly (and added the missing test that should've caught this), agreed on slack to do the refactor of registriesToSkip to happen here instead of digestResolver in a follow-on (just to keep this initial PR small)